### PR TITLE
[MetalPerformanceShaders] Neural Networks Update to Xcode 11

### PR DESCRIPTION
**Pull request by praeclarum**
This consolidates PRs #6932 and #6935 with a few extra additions.

It adds new functionality to the neural network components. This is still not the complete API for Xcode 11.

I have tested this code with a simple MNIST training app that mimics Apple's example from WWDC. It is available here: https://github.com/praeclarum/ImageRecognizer

I have also built a high level library using this code available here: https://github.com/praeclarum/MetalTensors

### CODE REVIEW

**Comment by praeclarum** [comment]
Thanks for the review, I'll get on it.

Can you help with these test failures? Not sure what causes, for example:

```
MetalPerformanceShaders.MPSMatrixRandom should re-expose MPSKernel::.ctor(Metal.IMTLDevice) 
```

**Comment by praeclarum** [comment]
I just added a new commit because I bound one of `RandomDistributionDescriptor`'s creation methods incorrectly.

**Comment by praeclarum** [comment]
I don't think I caused those last 40 test errors - lots of device deployment errors that I will chalk up to infrastructure.

**Comment by praeclarum** [comment]
@rolfbjarne Thanks for helping this along!

The tests will only work on hardware, I will have to investigate how to mark them as such.

File: src/metalperformanceshaders.cs [hunk 1]
```diff
@@ -6977,6 +7506,7 @@ interface MPSCnnGradientKernel {
 		void EncodeBatch (IMTLCommandBuffer commandBuffer, NSArray<MPSImage> sourceGradients, NSArray<MPSImage> sourceImages, NSArray<MPSState> gradientStates, NSArray<MPSImage> destinationGradients);
 	}
 
+	// [Introduced (PlatformName.MacCatalyst, 13, 0)]
```
[the remark this answers is not in the document]
Comment (praeclarum) [reply src/metalperformanceshaders.cs]: Good catch, that seems to be a mistake and I'll fix it (it should be there).

In general, there are **a lot** of APIs that haven't been updated with the `[Introduced (PlatformName.MacCatalyst, 13, 0)]` bit. I hope to do a second PR once these APIs make it through and will verify MacCatalyst.

**Comment by rolfbjarne** [comment]
> The tests will only work on hardware, I will have to investigate how to mark them as such.

Here's an example: https://github.com/xamarin/xamarin-macios/blob/d8468c621e6c0a12f4ba61eb8057842609a1e820/tests/monotouch-test/MetalPerformanceShaders/MPSStateBatchTests.cs#L28-L37

**Comment by dalexsoto** [comment]
Closing in favor of https://github.com/xamarin/xamarin-macios/pull/8844 thanks a lot Frank!

**Comment by praeclarum** [comment]
@dalexsoto It's annoying that I put so much effort into this and waited so long (8 months!) and don't get to be in the commit history. Why did this go so bad? What mistakes did I make in this PR that it was never accepted?

**Comment by dalexsoto** [comment]
Hello @praeclarum!

First of all I offer you an apology, it had nothing to do with you or your PR it was definitely us dropping the ball here since this was ready to be merged back in November. That said, I made sure you do appear in the commit history[1] but if you want to rebase it yourself on top of `d16-6-xcode11.6` and create your own PR please do and let me know so I can close #8844 in favor of yours.

Thanks a lot for your contribution, we really appreciate it and again I am very sorry for this oversight on our part, we'll make sure it doesn't happen again in the future.

[1]:
image

**Comment by praeclarum** [comment]
Thanks for the reply Alex. No, I don't need to stretch this PR out anymore and won't waste your time rebasing. I was just asking because this put a sour taste in my mouth for contributing and wanted to know what to do differently in the future if I dare again.

**Pull request closed without merging**